text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,254 @@
+are masked out prior to every load and store, it is possible to indirectly use a tagged pointer even
+without this. However, the way these kinds of architecture features tend to be applied involves
+every heap allocation coming with a tag, and then having to mask out every heap-allocated pointer
+would be putting a much heavier burden on the end-user. With this helper, we can have the library do
+the heavy lifting and the end-user does not even have to be aware that the pointers they are using
+happen to be tagged.
+
+For those one-off applications, an LLVM optimisation pass that optimises away applying the mask into
+NOPs on TBI-enabled platforms would be nice to have, although obviously that is out of the scope for
+this RFC.
+ 
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC adds one associated function to `core::ptr`:
+`pub unsafe fn assume_moved<T>(original: *mut T, new_address: usize) -> *mut T`
+
+```rust
+use core::ptr::assume_moved;
+
+let tag = 63;
+let new_addr = ptr as usize | tag << 56;
+let tagged_ptr = unsafe { assume_moved(ptr, new_addr) };
+```
+
+The purpose of this function is to indicate to the compiler that an allocation that used to be
+pointed to by a given pointer can now only be accessed by the new pointer with the provided new
+address. This is supposed to be semantically equivalent to a `realloc` from the untagged address to
+the tagged address, and conceptually similar to a `move` - it is no longer valid to access the
+allocation through the untagged pointer or any derived pointers. That being said, no actual
+moving is done - the underlying memory does not change, it only changes within the Rust memory
+model.
+
+Importantly, this is only designed to work on pointers to heap allocations. Trying to tag a
+stack-allocated variable in this way will almost definitely result in undefined behavior because
+the compiler will not be able to reason about whether writes to the tagged pointer can modify the
+value of the local variable.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As previously explained, the memory model we currently have is not fully compatible with memory
+tagging and tagged pointers. Setting the high bits of a pointer must be done with great care in
+order to avoid introducing undefined behavior, which could arise as a result of violating pointer
+aliasing rules - using two 'live' pointers which have different 64-bit addresses but do point to
+the same chunk of memory would weaken alias analysis and related optimisations.
+
+We can avoid this issue by assuming that a move from the untagged address to the tagged address has
+happened. To do so, we need the helper function to return a pointer with a brand new provenance,
+disjoint from the provenance of the original pointer. This can be achieved by the combination of
+using `inttoptr` and an inline asm block.
```

Comment:
Attaching a comment to this to record the concern that using `inttoptr` here makes no sense IMO. Also see the discussion [here](https://github.com/rust-lang/rfcs/pull/3700#discussion_r2093947303).

Reply:
That said -- why does the RFC even go into that much detail? The LLVM IR we emit is an implementation detail! The RFC should concern itself with the user-visible behavior and guarantees.

Maybe it should have a section on "implementation considerations" or so. But then it should *actually spell out the LLVM IR*, the description here is too vague to give me confidence that you and me have the same IR in our minds.

Reply:
Fair point, I've moved implementation considerations into their own section and specified that the function snippet is an _example_ implementation.